File: implementations/libraries/libcarlab.py
import calendar
import os
import pickle
import requests
import time

class libcarlab ():
    def __init__ (self, userid, required_info, output_info, save_filename, test=False):
        self.last_check_time = calendar.timegm(time.gmtime()) - 10000
        self.userid = userid
        self.test = test
        self.required_info = required_info
        self.baseurl = 'http://localhost:1234/packet/'
        self.fetch_url = '{base}list?information={info}&person={person}&sincetime={time}'
        self.push_url = '{base}upload?information={info}&person={person}'
        self.save_filename = save_filename

        if not os.path.exists(save_filename):
            ofile = open(save_filename, 'wb')
            pickle.dump({}, ofile)
            ofile.close()

    
    def check_new_info (self):
        # Make server call to /packet/list
        # if there are files, copy them over 
        # (stick to HTTP routing so we can put server scripts on a separate machine)
        # XXX do differently if testing
        new_data = {}

        for info in self.required_info:
            url = self.fetch_url.format(
                base=self.baseurl,
                info=info,
                person=self.userid,
                time=self.last_check_time,
            )
            
            results = requests.get(url)

            print(url)
            new_data[info] = results.json()

        self.last_check_time = calendar.timegm(time.gmtime())
        return new_data
    
    def output_new_info (self, info, value):
        # XXX do differently for test

        # save to file basically. Then we'll try uploading separately
        existingData = pickle.load(open(self.save_filename, 'rb'))
        existingData.setdefault(info, [])
        existingData[info]append(value)
        ofile = open(self.save_filename, 'wb')
        pickle.dump(existingData, ofile)
        ofile.close()
    
    
    def upload_data (self):
        for info in self.output_info:
            existingData = pickle.load(open(self.save_filename, 'rb'))

            if info in existingData:
                result = requests.post(self.push_url.format(
                    base=self.baseurl,
                    info=info,
                    person=self.userid,
                ), { 'message': existingData[info] })
            
                if result.status_code == 200:
                    del existingData[info]
                    ofile = open(self.save_filename, 'wb')
                    pickle.dump(existingData, ofile)
                    ofile.close()

File: implementations/libraries/Libcarlab.js
export class Libcarlab {
  constructor (userid, requiredInfo, test, storageHandler) {
    this.lastCheckTime = Math.round(new Date().getTime() / 1000)
    this.userid = userid
    this.test = test
    this.requiredInfo = requiredInfo

    this.localStorage = {}

    this.baseurl = 'http://localhost:1234/packet/'
    this.fetchUrl = (info) => `${this.baseurl}list?information=${info}&person=${this.userid}&sincetime=${this.lastCheckTime}`
    this.pushUrl = (info) => `${this.baseurl}upload?information=${info}&person=${this.userid}`

    this.uploadTimer = false;
    this.uploadEvery = 30*1000; // 30 seconds
    this.storageHandler = storageHandler
  }
  
  checkNewInfo (callback) {
    // Get new data
    // check with local storage first to only get relevant data
    for (let info of this.requiredInfo) 
        fetch(this.fetchUrl(info))
            .then(res => res.json())
            .then(data => callback(info, data));
    
    // XXX This should only be updated IF we successfully made a call. Otherwise info will be lost
    // This could also be the time of the last data we received -- that might be better to avoid time-off-sync-related errors
    this.lastCheckTime = Math.round(new Date().getTime() / 1000)
  }

  outputNewInfo (info, message) {
    console.log('Sending to ', this.pushUrl(info), 'value', info, message);
    
    if (!this.localStorage.hasOwnProperty(info))
      this.localStorage[info] = []
    this.localStorage[info].push(message)
  }

  
  scheduleUploads () {
    if (this.uploadTimer)
      return
    
    this.uploadTimer = setInterval(() => this.performUpload, this.uploadEvery)
  }

  async performUpload () {
    if (Object.keys(this.localStorage).length === 0)
    return;
    
    // Save everything to disk (in case app crashes)
    for (let key in this.localStorage) 
      if (this.localStorage.hasOwnProperty(key)) {
        await this.storageHandler.saveOrUpdateData(key, this.localStorage[key]);
        delete this.localStorage[key];
      }
    

    // Read from disk
    let allOutput = await this.storageHandler.loadAllData()
      
    // Upload each information
    for (let info in allOutput) {
      if (allOutput.hasOwnProperty(info)) {
        fetch(this.push_url(info), {
          method: 'post', mode: 'cors', cache: 'no-cache',
          headers: { 'Content-type': 'application/json' },
          body: JSON.stringify({'message': allOutput[info] })})
          .then(res => {
            // Success
            await this.storageHandler.clearData(info);
          })
      }
    }
  }

  unscheduleUploads () {
    if (this.uploadTimer) {
      clearInterval(this.uploadTimer)
      this.uploadTimer = false;
    }
  }
}